### PR TITLE
Add setting to disable user features when user login type is not plain

### PR DESCRIPTION
**Pull request by jackHay22**
## Changes
- Adds setting `EXTERNAL_USER_DISABLE_FEATURES` to disable any supported user features when login type is not plain
  - In general, this is necessary for SSO implementations to avoid inconsistencies between the external account management and the linked account
- Adds helper functions to encourage correct use

### CODE REVIEW

**Comment by lunny** [comment]
But I don't think `password` is meaningful for external users.

**Comment by jackHay22** [comment]
> But I don't think `password` is meaningful for external users.

Do you mean that changing a password should be disabled? I agree but I'm not sure if that's directly related to the changes here.

**Comment by lunny** [comment]
What if the user system needs to be refactored in future that a user will always be a plain one and can linked to external account?

**Comment by jackHay22** [comment]
> What if the user system needs to be refactored in future that a user will always be a plain one and can linked to external account?

So checking `LoginType` wouldn't be the best way to check if the account is linked to an external provider? Should we instead check for an `ExternalLoginUser`?

**Comment by lunny** [comment]
> > What if the user system needs to be refactored in future that a user will always be a plain one and can linked to external account?
> 
> So checking `LoginType` wouldn't be the best way to check if the account is linked to an external provider? Should we instead check for an `ExternalLoginUser`?

Currently I think check `LoginType` is OK because not every external accounts have been saved into `ExternalLoginUser`. But we need to have a comment here, once all external login users have been stored, then we can check the table.

File: custom/conf/app.example.ini [hunk 1]
```diff
@@ -1485,6 +1485,8 @@ LEVEL = Info
 ;; - manage_ssh_keys: a user cannot configure ssh keys
 ;; - manage_gpg_keys: a user cannot configure gpg keys
 ;USER_DISABLED_FEATURES =
+;; Whether to disable all user features if the user has an external login type
+;;EXTERNAL_USER_DISABLE_FEATURES = false
```
Comment (lunny) [inline custom/conf/app.example.ini]:
```suggestion
;;EXTERNAL_USER_DISABLE_ALL_FEATURES = false
```

**Review by lunny — APPROVED** [review]

**Comment by lunny** [comment]
And please add documents and tests.

File: docs/content/administration/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (lunny) [inline docs/content/administration/config-cheat-sheet.en-us.md]: unnecessary change.

**Comment by jackHay22** [comment]
I think it would be better to change `EXTERNAL_USER_DISABLE_ALL_FEATURES` to `EXTERNAL_USER_DISABLE_FEATURES` and allow selective inclusion.

File: custom/conf/app.example.ini [hunk 3]
```diff
@@ -1485,6 +1485,8 @@ LEVEL = Info
 ;; - manage_ssh_keys: a user cannot configure ssh keys
 ;; - manage_gpg_keys: a user cannot configure gpg keys
 ;USER_DISABLED_FEATURES =
+;; Whether to disable all user features if the user has an external login type
```
Comment (silverwind) [inline custom/conf/app.example.ini]: Copy full description in here as well, wrapped.

File: models/user/user.go [hunk 4]
```diff
@@ -1232,3 +1232,21 @@ func GetOrderByName() string {
 	}
 	return "name"
 }
+
+// FeatureDisabledWithLoginType checks if a user feature is disabled, taking into account the login type of the
+// user if applicable
+func FeatureDisabledWithLoginType(user *User, feature string) bool {
```
Comment (silverwind) [inline models/user/user.go]: Rename to `IsFeatureDisabledWithLoginType`, boolean-returning functions should always be prefixed with `Is` in my book.

File: custom/conf/app.example.ini [hunk 5]
```diff
@@ -1485,6 +1485,11 @@ LEVEL = Info
 ;; - manage_ssh_keys: a user cannot configure ssh keys
 ;; - manage_gpg_keys: a user cannot configure gpg keys
 ;USER_DISABLED_FEATURES =
+;; Disabled features if the user has an external login type. Can be the same options as for USER_DISABLED_FEATURES.
+;; - deletion: a user cannot delete their own account
+;; - manage_ssh_keys: a user cannot configure ssh keys
+;; - manage_gpg_keys: a user cannot configure gpg keys
+;;EXTERNAL_USER_DISABLE_FEATURES =
```
Comment (silverwind) [inline custom/conf/app.example.ini:1492]: What is the syntax of this? Comma-separated? If so, start sentence with `Comma-separated list of ...`

File: docs/content/administration/config-cheat-sheet.en-us.md [hunk 6]
```diff
@@ -522,6 +522,10 @@ And the following unique queues:
   - `deletion`: User cannot delete their own account.
   - `manage_ssh_keys`: User cannot configure ssh keys.
   - `manage_gpg_keys`: User cannot configure gpg keys.
+- `EXTERNAL_USER_DISABLE_FEATURES`: **_empty_**: Disabled features ONLY if the user has an external login type (eg. LDAP, Oauth, etc.), could be `deletion`, `manage_ssh_keys`, `manage_gpg_keys` and more features can be added in future. This setting is independent from `USER_DISABLED_FEATURES` and supplements its behavior.
```
Comment (silverwind) [inline docs/content/administration/config-cheat-sheet.en-us.md]: Drop the "and more features can be added in future" part. Just explicitely list the valid values.

After these doc changes are done, copy the same sentence from Markdown into the ini imho. No need to maintain 2 different versions.

**Review by techknowlogick — APPROVED** [review]

**Review by silverwind — APPROVED** [review]
